pkg/internalapis/featurestates/featurestates.go

```diff
@@ -0,0 +1,399 @@
+	}
+	cnsCsiSvFeatureStates := &featurestatesv1alpha1.CnsCsiSvFeatureStates{
+		ObjectMeta: metav1.ObjectMeta{Name: SVFeatureStateCRName, Namespace: namespace},
+		Spec: featurestatesv1alpha1.CnsCsiSvFeatureStatesSpec{
+			FeatureStates: featureStates,
+		},
+	}
+	err = controllerRuntimeClient.Create(ctx, cnsCsiSvFeatureStates)
+	if err != nil {
+		log.Errorf("failed to create cnsCsiSvFeatureStates CR instance in the namespace: %q, Err: %v", cnsCsiSvFeatureStates.Namespace, err)
+		pendingCRUpdatesObj.pendingCRUpdate[namespace] = cnsCsiSvFeatureStates
+	}
+	return nil
+}
+
+// configMapUpdated is the call back function for config-map informer
+func configMapUpdated(oldObj, newObj interface{}) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	ctx = logger.NewContextWithLogger(ctx)
+	log := logger.GetLogger(ctx)
+
+	newfssConfigMap, ok := newObj.(*v1.ConfigMap)
+	if newfssConfigMap == nil || !ok {
+		log.Warnf("configMapUpdated: unrecognized old object %+v", newObj)
+		return
+	}
+	oldfssConfigMap, ok := oldObj.(*v1.ConfigMap)
+	if oldfssConfigMap == nil || !ok {
+		log.Warnf("configMapUpdated: unrecognized new object %+v", newObj)
+		return
+	}
+
+	if newfssConfigMap.Name == supervisorFeatureStatConfigMapName &&
+		newfssConfigMap.Namespace == supervisorFeatureStateConfigMapNamespace {
+		if reflect.DeepEqual(newfssConfigMap.Data, oldfssConfigMap.Data) {
+			return
+		}
+		var err error
+		var featureStates []featurestatesv1alpha1.FeatureState
+		for feature, state := range newfssConfigMap.Data {
+			var featureState featurestatesv1alpha1.FeatureState
+			featureState.Name = feature
+			featureState.Enabled, err = strconv.ParseBool(state)
+			if err != nil {
+				log.Errorf("failed to parse feature state value: %v for feature: %q", state, feature)
+				return
+			}
+			featureStates = append(featureStates, featureState)
+		}
+		for {
+			// if failed to create or update CR in the namespace, then error is not returned
+			// for such cases, CR update will be enqueued to pendingCRUpdatesObj
+			err := updateFeatureStatesCRInAllWorkLoadNamespaces(ctx, featureStates)
+			if err == nil {
+				log.Infof("updated feature states: %v in the all workload namespaces", featureStates)
+				break
+			} else {
+				log.Errorf("failed to update feature states CR in the all workload namespaces. err: %v", err)
+			}
+			time.Sleep(2 * time.Second)
+		}
+	}
+}
+
+// configMapAdded adds feature state switch values from configmap that has been created on K8s cluster
+func namespaceAdded(obj interface{}) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	ctx = logger.NewContextWithLogger(ctx)
+	log := logger.GetLogger(ctx)
+
+	namespace, ok := obj.(*v1.Namespace)
+	if namespace == nil || !ok {
+		log.Warnf("namespaceAdded: unrecognized object %+v", obj)
+		return
+	}
+	if _, ok = namespace.Labels[WorkLoadNamespaceLabelKey]; ok {
+		log.Infof("observed new workload namespace: %v", namespace.Name)
+		for {
```

Comment:
Why do we have an infinite loop?

Reply:
this is for handling errors other than create CR, when the latest feature switch state is not retrieved or we fail to parse feature switch state. added retry in the loop here, so I do not miss the namespace add event.